libgalois/include/katana/ParallelSTL.h

```diff
@@ -391,6 +393,102 @@ partial_sum(InputIt first, InputIt last, OutputIt d_first) {
+OutputIt
+transform(
+    InputIt first, InputIt last, OutputIt d_first, UnaryOperation unary_op) {
+  // This will only work for random access iterators
+  using input_category =
+      typename std::iterator_traits<InputIt>::iterator_category;
+  using output_category =
+      typename std::iterator_traits<OutputIt>::iterator_category;
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, input_category>);
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, output_category>);
+
+  using diff_type = typename std::iterator_traits<InputIt>::difference_type;
+
+  on_each([&](unsigned tid, unsigned total) {
+    auto [begin, end] = block_range(first, last, tid, total);
+    diff_type offset = std::distance(first, begin);
+    std::transform(begin, end, d_first + offset, unary_op);
+  });
+
+  return d_first + std::distance(first, last);
+}
+
+template <class InputIt, class OutputIt>
+OutputIt
+copy(InputIt first, InputIt last, OutputIt d_first) {
+  // This will only work for random access iterators
+  using input_category =
+      typename std::iterator_traits<InputIt>::iterator_category;
+  using output_category =
+      typename std::iterator_traits<OutputIt>::iterator_category;
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, input_category>);
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, output_category>);
+
+  using diff_type = typename std::iterator_traits<InputIt>::difference_type;
+
+  on_each([&](unsigned tid, unsigned total) {
+    auto [begin, end] = block_range(first, last, tid, total);
+    diff_type offset = std::distance(first, begin);
+    std::copy(begin, end, d_first + offset);
+  });
+
+  return d_first + std::distance(first, last);
+}
+
+template <class InputIt, class OutputIt, class UnaryPredicate>
+OutputIt
+copy_if(InputIt first, InputIt last, OutputIt d_first, UnaryPredicate pred) {
+  // This will only work for random access iterators
+  using input_category =
+      typename std::iterator_traits<InputIt>::iterator_category;
+  using output_category =
+      typename std::iterator_traits<OutputIt>::iterator_category;
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, input_category>);
+  static_assert(
+      std::is_base_of_v<std::random_access_iterator_tag, output_category>);
+
+  using diff_type = typename std::iterator_traits<InputIt>::difference_type;
+
+  // first on_each, set ranges
+  uint32_t num_threads = getActiveThreads();
+  std::vector<diff_type> prefix_sum(num_threads);
+  katana::on_each([&](unsigned tid, unsigned total) {
+    auto [begin, end] = katana::block_range(first, last, tid, total);
+    diff_type count = 0;
+    for (; first != last; ++first) {
+      if (pred(first)) {
+        count++;
+      }
+    }
+    prefix_sum[tid] = count;
+  });
+
+  // calculate prefix sums
+  for (size_t tid = 1; tid < prefix_sum.size(); ++tid) {
```

Comment:
BTW, https://en.cppreference.com/w/cpp/algorithm/partial_sum

(an explicit loop is fine as well)

small preference for avoiding `prefix_sum.size()` in the loop condition though: `for (size_t tid = ...., n = ...; tid < n; ++tid) ....`